### DISCUSSION TITLE
Multiple projects in local Supabase

### DISCUSSION THREAD

**Discussion by martwozniak** [supabase/supabase#5968] [Questions] [answered]
Hi,

When working with local Supabase (via Docker), I didn't see the option to add multiple projects. The cloud version has this possibility. 

Do you plan to add the ability to create multiple projects on Supabase locally? 

Or should I create multiple instances of supabase, each under a separate project that uses it?

⚠️ **Important:**
Answer here (i cant change pinned comment): https://github.com/orgs/supabase/discussions/5968#discussioncomment-9059734

**Comment by GaryAustin1** [comment]
This was the last thing I’ve seen on this…
https://github.com/supabase/supabase/issues/4440

**Comment by prodkt** [comment]
With multiple projects inside a local supabase out of scope and not an option moving forward will there be any complications running 2 or more local supabase stacks with docker running through the regular install/setup process?

I'm trying to confirm before any dev time is spent on this. My curiosity on how this works limited to 1, and having to run multiple containers is in the container names, ports, db(db) user(postgres) exactly the same when composing up? We foresee in the near future needing the abstraction but also the ease to communicate back and fourth how we'd have them in production.

Depending on the answer to the above question the following context may help. I'd prefer not in anyway manipulate the supabase clones if we must have several. The hope would be they all remain true to your deploy without alteration to avoid any migration or deployment issues.

@GaryAustin1 any help would be appreciated answering the above 👍 Thank you!

**Comment by liyangguang** [comment]
You can achieve it by running multiple instances of supabase docker image.

It's not super convenient out of the box, but you can write a script to easily automate this.
- make a copy of the `docker` folder in supabase repo
- (assume one local instances of supabase is running / had run before, thus having name conflicts) rename all the supabase docker images for your project. like `docker rename supabase-auto project-foo-supabase-auto`, there are about 10 containers. Use `docker ps -a` to find them, they all have `supabase` in the name.
- Assign ports. In `.env` (I assume `/volumnes/api/kong.yml` as well), update all the port numbers so it doesn't conflict with the other project. (My quick and dirty way is run the file through `fileString.replace(/(\d{4,})/g, (match) => Number(match) + 100);` to add 100 to all the port numbers)

Now you can have both instances running, one on :3000, with API on :8000. and another on :3100, with API on :8100, totally separate database and everything.

(I only tested the basic usage, haven't fully tested if everything is working)

**Reply by ChuckJonas** [reply]
any update on how this worked out for you?  

When doing this, what happens to the `supabase start/stop` commands?  Do they continue to work for each project?

**Reply by liyangguang** [reply]
No need to start stop. They are all just separate containers, can run simultaneously on the same machine, as long as you have different docker name and ports as I noted above.
Each of the instance works exactly like what's described in the official documentation.

**Comment by MaximusMcCann** [comment]
Would love to be able to specify the `config.toml` file to use.  Would be ideal when running some local integration tests 🙏 .

```
supabase start --config 'config-test.toml'
```

**Comment by code-leen** [comment]
I've found that updating the `project_id` in the `config.toml` (along with the port number values in the file) makes it possible to run multiple Supabase projects locally!

https://supabase.com/docs/guides/cli/config

**Reply by znowfox** [reply]
Could you please tell me more ? new overall folder, diff docker images, selectable projects ? Please le met know.

**Reply by znowfox** [reply]
time waster. I'll post a bash script to automate all of this BS.

**Reply by Geczy** [reply]
@znowfox did you post the script yet?

**Reply by znowfox** [reply]
> @znowfox did you post the script yet? 

Pm

**Reply by Geczy** [reply]
> > @znowfox did you post the script yet? 
> 
> Pm

What

**Reply by martwozniak** [reply]
I know that you put a lot of effort into the creation of this automation. If you can please share it with the community, I believe that a lot of people have the same problem. We will appreciate it!

**Reply by wave-light** [reply]
Do we need a bash script? From my understanding of what @code-leen said, we just need to ensure that the projects have different project ids and port numbers assigned in config.toml, then we should be able to run both projects simultaneously.

**Reply by code-leen** [reply, marked as answer]
@znowfox @Geczy @martwozniak 
I updated the port numbers and the `project_id` in the `config.toml`. These values _must_ be different between your projects so they do not conflict on your local machine. The `config.toml` can be found at the '/supabase' directory in your project. This solution works for my use-case with having two different database instances utilizing Supabase, and I can see the two different sets of containers when viewing it in the Docker dashboard.

After making changes to these files, [the docs](https://supabase.com/docs/guides/cli/config) mention you will need to run `supabase stop` and `supabase start` for it to work.

Here is an example of how the `config.toml` files will look.

**Project 1**:
```
# A string used to distinguish different Supabase projects on the same host. Defaults to the working
# directory name when running `supabase init`.
project_id = "project-1"

[api]
# Port to use for the API URL.
port = 54321
# Schemas to expose in your API. Tables, views and stored procedures in this schema will get API
# endpoints. public and storage are always included.
schemas = ["public", "storage", "graphql_public"]
# Extra schemas to add to the search_path of every request. public is always included.
extra_search_path = ["public", "extensions"]
# The maximum number of rows returns from a view, table, or stored procedure. Limits payload size
# for accidental or malicious requests.
max_rows = 1000

[db]
# Port to use for the local database URL.
port = 54322
# The database major version to use. This has to be the same as your remote database's. Run `SHOW
# server_version;` on the remote database to check.
major_version = 15

[studio]
# Port to use for Supabase Studio.
port = 54323

# Email testing server. Emails sent with the local dev setup are not actually sent - rather, they
# are monitored, and you can view the emails that would have been sent from the web interface.
[inbucket]
# Port to use for the email testing server web interface.
port = 54324
smtp_port = 54325
pop3_port = 54326

# the rest of the config.toml file....
```

**Project 2**:
```
# A string used to distinguish different Supabase projects on the same host. Defaults to the working
# directory name when running `supabase init`.
project_id = "project-2"

[api]
# Port to use for the API URL.
port = 64321
# Schemas to expose in your API. Tables, views and stored procedures in this schema will get API
# endpoints. public and storage are always included.
schemas = ["public", "storage", "graphql_public"]
# Extra schemas to add to the search_path of every request. public is always included.
extra_search_path = ["public", "extensions"]
# The maximum number of rows returns from a view, table, or stored procedure. Limits payload size
# for accidental or malicious requests.
max_rows = 1000

[db]
# Port to use for the local database URL.
port = 64322
# The database major version to use. This has to be the same as your remote database's. Run `SHOW
# server_version;` on the remote database to check.
major_version = 15

[studio]
# Port to use for Supabase Studio.
port = 64323

# Email testing server. Emails sent with the local dev setup are not actually sent - rather, they
# are monitored, and you can view the emails that would have been sent from the web interface.
[inbucket]
# Port to use for the email testing server web interface.
port = 64324
smtp_port = 64325
pop3_port = 64326

# the rest of the config.toml file....
```

Let me know if you have any more questions! Happy to clarify.

**Reply by Geczy** [reply]
oh cool, so this way you're not duplicating the supabase services, you're only duplicating a postgres db?

**Reply by wave-light** [reply]
@Geczy No, you're not duplicating anything; because of the different project IDs in config.toml, you get completely separate services and DB, in different docker containers.

**Reply by znowfox** [reply]
You're definitely duplicating everything, which is fine!

**Reply by Geczy** [reply]
if we're just duplicating, then i'd like to suggest using coolify to automate deploying a supabase service. works out of the box already

**Reply by martwozniak** [reply]
@code-leen looks solid, I marked your answer as an answer to the issue. Great work i appreciate it  🙌

**Reply by martwozniak** [reply]
@Geczy can you post solution that uses coolify? I think that somebody can be interested, conversations like that are great for community

**Reply by znowfox** [reply]
and realtime don't work that way, btw

**Reply by Geczy** [reply]
https://github.com/coollabsio/coolify

they have a supabase template you can just run to get setup

**Reply by JoyShaheb** [reply]
don't know why but when i run `supabase start` i get the error `supabase_vector_supabase container is not ready: unhealthy` any reason i am doing wrong ? i just disabled the github thing on the `config.toml` file 
![image](https://github.com/user-attachments/assets/df7d8ac9-37e7-432a-8063-d1a153f45c50)

**Reply by 2-fly-4-ai** [reply]
> don't know why but when i run `supabase start` i get the error `supabase_vector_supabase container is not ready: unhealthy` any reason i am doing wrong ? i just disabled the github thing on the `config.toml` file ![image](https://private-user-images.githubusercontent.com/61899866/359349123-df7d8ac9-37e7-432a-8063-d1a153f45c50.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc2MDQ1NDAsIm5iZiI6MTcyNzYwNDI0MCwicGF0aCI6Ii82MTg5OTg2Ni8zNTkzNDkxMjMtZGY3ZDhhYzktMzdlNy00MzJhLTgwNjMtZDFhMTUzZjQ1YzUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI5VDEwMDQwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzNTg1ZDZjOTBhNmNiMjg2NTY4OWU2MzMwNGE2ODA2NjRmYzdiNTI0MGZhZTY1ZWNmNzcwNjAzZDhkNDg4YzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yqm6Szhe9mgAGziFpJQVkci_3Jy2kGAeCgiM1Pn3F00)

I get the exact same issue. Did you manage to solve this?

**Reply by guyrandalf** [reply]
I does not seem to matter. Duplicate or not, it works based on projects which is totally fine!

**Reply by dungpv04** [reply]
> @Geczy No, you're not duplicating anything; because of the different project IDs in config.toml, you get completely separate services and DB, in different docker containers.

So you mean this solution still duplicates Supabase self-hosted docker containers? Usually, I changed the container names, ports in its docker-compose.yml and it will run another compose with duplicate supabase containers but different names and ports, which mean it still consumes really much resource.

**Comment by 100tomtomas100** [comment]
@liyangguang @code-leen do these ways work with docker ? 

@liyangguang  I am getting a lot of errors trying to change all the services ports as there are a lot of settings in the volumes with preset ports.
@code-leen  do you use one config.toml  or you have the setup of supabase twice and then you configure config.toml differently for each setup? and how do you proceed after you arrange config.toml? 

I am sorry for asking stupid questions but I am new to docker and supabase and there is not much info about setting up multiple databases in the same server with supabase.

**Reply by code-leen** [reply]
@100tomtomas100 No worries at all, these are great questions! 

Yes, it works with the Docker set-up that Supabase runs when you start it.

For each project/repo I have that utilizes its own database with Supabase, it has its own config.toml file. For example, Project A has its own config.toml with its own unique set of ports. Project B also has its own config.toml with a different unique set of ports.

By modifying the `project_id` and the ports between each project in the  config.toml, it is possible to run multiple Supabase projects on the same machine.

Reference: https://supabase.com/docs/guides/cli/config#project_id

**Reply by code-leen** [reply]
@100tomtomas100 To answer your second question for how to proceed after these modifications, you will need to restart your project using `supabase stop` and `supabase start` according to the [docs](https://supabase.com/docs/guides/cli/config).

**Reply by 100tomtomas100** [reply]
@code-leen  thank you for the answer 

For example I have a project bar and I need 2 databases there
How do I create 2 database for this one project?
bar/supabase-1 and bar/supabase-2
Should supabase-1 and supabase-2 be new supabase setups or they should be made somehow from 1 setup? 
Change config.toml  (project name and ports) - this is the only thing clear
Do I need to install supabase cli to run supabase start and stop? It is not enough to run docker compose down and docker compose up?  
Even though I am changing ports in config.toml the services ports and services names are not clashing? 

sorry for the many questions :(

**Reply by 2-fly-4-ai** [reply]
I've tried to follow the instructions, I keep running into errors.

Before
```

# A string used to distinguish different Supabase projects on the same host. Defaults to the
# working directory name when running `supabase init`.
project_id = "fiji-tax-pro"

[api]
enabled = true
# Port to use for the API URL.
port = 54321
# Schemas to expose in your API. Tables, views and stored procedures in this schema will get API
# endpoints. `public` is always included.
schemas = ["public", "graphql_public"]
# Extra schemas to add to the search_path of every request. `public` is always included.
extra_search_path = ["public", "extensions"]
# The maximum number of rows returns from a view, table, or stored procedure. Limits payload size
# for accidental or malicious requests.
max_rows = 1000

[api.tls]
enabled = false

[db]
# Port to use for the local database URL.
port = 54322
# Port used by db diff command to initialize the shadow database.
shadow_port = 54320
# The database major version to use. This has to be the same as your remote database's. Run `SHOW
# server_version;` on the remote database to check.
major_version = 15

[db.pooler]
enabled = false
# Port to use for the local connection pooler.
port = 54329
# Specifies when a server connection can be reused by other clients.
# Configure one of the supported pooler modes: `transaction`, `session`.
pool_mode = "transaction"
# How many server connections to allow per user/database pair.
default_pool_size = 20
# Maximum number of client connections allowed.
max_client_conn = 100

[realtime]
enabled = true
# Bind realtime via either IPv4 or IPv6. (default: IPv4)
# ip_version = "IPv6"
# The maximum length in bytes of HTTP request headers. (default: 4096)
# max_header_length = 4096

[studio]
enabled = true
# Port to use for Supabase Studio.
port = 54323
# External URL of the API server that frontend connects to.
api_url = "http://127.0.0.1"
# OpenAI API Key to use for Supabase AI in the Supabase Studio.
openai_api_key = "env(OPENAI_API_KEY)"

# Email testing server. Emails sent with the local dev setup are not actually sent - rather, they
# are monitored, and you can view the emails that would have been sent from the web interface.
[inbucket]
enabled = true
# Port to use for the email testing server web interface.
port = 54324
# Uncomment to expose additional ports for testing user applications that send emails.
# smtp_port = 54325
# pop3_port = 54326

[storage]
enabled = true
# The maximum file size allowed (e.g. "5MB", "500KB").
file_size_limit = "50MiB"

[storage.image_transformation]
enabled = true

# Uncomment to configure local storage buckets
# [storage.buckets.images]
# public = false
# file_size_limit = "50MiB"
# allowed_mime_types = ["image/png", "image/jpeg"]
# objects_path = "./images"

[auth]
enabled = true
# The base URL of your website. Used as an allow-list for redirects and for constructing URLs used
# in emails.
site_url = "http://127.0.0.1:3000"
# A list of *exact* URLs that auth providers are permitted to redirect to post authentication.
additional_redirect_urls = ["https://127.0.0.1:3000"]
# How long tokens are valid for, in seconds. Defaults to 3600 (1 hour), maximum 604,800 (1 week).
jwt_expiry = 3600
# If disabled, the refresh token will never expire.
enable_refresh_token_rotation = true
# Allows refresh tokens to be reused after expiry, up to the specified interval in seconds.
# Requires enable_refresh_token_rotation = true.
refresh_token_reuse_interval = 10
# Allow/disallow new user signups to your project.
enable_signup = true
# Allow/disallow anonymous sign-ins to your project.
enable_anonymous_sign_ins = false
# Allow/disallow testing manual linking of accounts
enable_manual_linking = false

[auth.email]
# Allow/disallow new user signups via email to your project.
enable_signup = true
# If enabled, a user will be required to confirm any email change on both the old, and new email
# addresses. If disabled, only the new email is required to confirm.
double_confirm_changes = true
# If enabled, users need to confirm their email address before signing in.
enable_confirmations = false
# If enabled, users will need to reauthenticate or have logged in recently to change their password.
secure_password_change = false
# Controls the minimum amount of time that must pass before sending another signup confirmation or password reset email.
max_frequency = "1s"

# Use a production-ready SMTP server
# [auth.email.smtp]
# host = "smtp.sendgrid.net"
# port = 587
# user = "apikey"
# pass = "env(SENDGRID_API_KEY)"
# admin_email = "admin@email.com"
# sender_name = "Admin"

# Uncomment to customize email template
# [auth.email.template.invite]
# subject = "You have been invited"
# content_path = "./supabase/templates/invite.html"

[auth.sms]
# Allow/disallow new user signups via SMS to your project.
enable_signup = true
# If enabled, users need to confirm their phone number before signing in.
enable_confirmations = false
# Template for sending OTP to users
template = "Your code is {{ .Code }} ."
# Controls the minimum amount of time that must pass before sending another sms otp.
max_frequency = "5s"

# Use pre-defined map of phone number to OTP for testing.
# [auth.sms.test_otp]
# 4152127777 = "123456"

# Configure logged in session timeouts.
# [auth.sessions]
# Force log out after the specified duration.
# timebox = "24h"
# Force log out if the user has been inactive longer than the specified duration.
# inactivity_timeout = "8h"

# This hook runs before a token is issued and allows you to add additional claims based on the authentication method used.
# [auth.hook.custom_access_token]
# enabled = true
# uri = "pg-functions://<database>/<schema>/<hook_name>"

# Configure one of the supported SMS providers: `twilio`, `twilio_verify`, `messagebird`, `textlocal`, `vonage`.
[auth.sms.twilio]
enabled = false
account_sid = ""
message_service_sid = ""
# DO NOT commit your Twilio auth token to git. Use environment variable substitution instead:
auth_token = "env(SUPABASE_AUTH_SMS_TWILIO_AUTH_TOKEN)"

[auth.mfa]
# Control how many MFA factors can be enrolled at once per user.
max_enrolled_factors = 10

# Control use of MFA via App Authenticator (TOTP)
[auth.mfa.totp]
enroll_enabled = true
verify_enabled = true

# Configure Multi-factor-authentication via Phone Messaging
# [auth.mfa.phone]
# enroll_enabled = true
# verify_enabled = true
# otp_length = 6
# template = "Your code is {{ .Code }} ."
# max_frequency = "10s"

# Use an external OAuth provider. The full list of providers are: `apple`, `azure`, `bitbucket`,
# `discord`, `facebook`, `github`, `gitlab`, `google`, `keycloak`, `linkedin_oidc`, `notion`, `twitch`,
# `twitter`, `slack`, `spotify`, `workos`, `zoom`.
[auth.external.apple]
enabled = false
client_id = ""
# DO NOT commit your OAuth provider secret to git. Use environment variable substitution instead:
secret = "env(SUPABASE_AUTH_EXTERNAL_APPLE_SECRET)"
# Overrides the default auth redirectUrl.
redirect_uri = ""
# Overrides the default auth provider URL. Used to support self-hosted gitlab, single-tenant Azure,
# or any other third-party OIDC providers.
url = ""
# If enabled, the nonce check will be skipped. Required for local sign in with Google auth.
skip_nonce_check = false

# Use Firebase Auth as a third-party provider alongside Supabase Auth.
[auth.third_party.firebase]
enabled = false
# project_id = "my-firebase-project"

# Use Auth0 as a third-party provider alongside Supabase Auth.
[auth.third_party.auth0]
enabled = false
# tenant = "my-auth0-tenant"
# tenant_region = "us"

# Use AWS Cognito (Amplify) as a third-party provider alongside Supabase Auth.
[auth.third_party.aws_cognito]
enabled = false
# user_pool_id = "my-user-pool-id"
# user_pool_region = "us-east-1"

[edge_runtime]
enabled = true
# Configure one of the supported request policies: `oneshot`, `per_worker`.
# Use `oneshot` for hot reload, or `per_worker` for load testing.
policy = "oneshot"
inspector_port = 8083

[analytics]
enabled = true
port = 54327
# Configure one of the supported backends: `postgres`, `bigquery`.
backend = "postgres"

# Experimental features may be deprecated any time
[experimental]
# Configures Postgres storage engine to use OrioleDB (S3)
orioledb_version = ""
# Configures S3 bucket URL, eg. <bucket_name>.s3-<region>.amazonaws.com
s3_host = "env(S3_HOST)"
# Configures S3 bucket region, eg. us-east-1
s3_region = "env(S3_REGION)"
# Configures AWS_ACCESS_KEY_ID for S3 bucket
s3_access_key = "env(S3_ACCESS_KEY)"
# Configures AWS_SECRET_ACCESS_KEY for S3 bucket
s3_secret_key = "env(S3_SECRET_KEY)"
```
After

``` # A string used to distinguish different Supabase projects on the same host. Defaults to the
# working directory name when running `supabase init`.
project_id = "fiji-tax-pro"

[api]
enabled = true
# Port to use for the API URL.
port = 64321
# Schemas to expose in your API. Tables, views and stored procedures in this schema will get API
# endpoints. `public` is always included.
schemas = ["public", "graphql_public"]
# Extra schemas to add to the search_path of every request. `public` is always included.
extra_search_path = ["public", "extensions"]
# The maximum number of rows returns from a view, table, or stored procedure. Limits payload size
# for accidental or malicious requests.
max_rows = 1000

[api.tls]
enabled = false

[db]
# Port to use for the local database URL.
port = 64322
# Port used by db diff command to initialize the shadow database.
shadow_port = 64320
# The database major version to use. This has to be the same as your remote database's. Run `SHOW
# server_version;` on the remote database to check.
major_version = 15

[db.pooler]
enabled = false
# Port to use for the local connection pooler.
port = 64329
# Specifies when a server connection can be reused by other clients.
# Configure one of the supported pooler modes: `transaction`, `session`.
pool_mode = "transaction"
# How many server connections to allow per user/database pair.
default_pool_size = 20
# Maximum number of client connections allowed.
max_client_conn = 100

[realtime]
enabled = true
# Bind realtime via either IPv4 or IPv6. (default: IPv4)
# ip_version = "IPv6"
# The maximum length in bytes of HTTP request headers. (default: 4096)
# max_header_length = 4096

[studio]
enabled = true
# Port to use for Supabase Studio.
port = 64323
# External URL of the API server that frontend connects to.
api_url = "http://127.0.0.1"
# OpenAI API Key to use for Supabase AI in the Supabase Studio.
openai_api_key = "env(OPENAI_API_KEY)"

# Email testing server. Emails sent with the local dev setup are not actually sent - rather, they
# are monitored, and you can view the emails that would have been sent from the web interface.
[inbucket]
enabled = true
# Port to use for the email testing server web interface.
port = 64324
# Uncomment to expose additional ports for testing user applications that send emails.
# smtp_port = 64325
# pop3_port = 64326

[storage]
enabled = true
# The maximum file size allowed (e.g. "5MB", "500KB").
file_size_limit = "50MiB"

[storage.image_transformation]
enabled = true

# Uncomment to configure local storage buckets
# [storage.buckets.images]
# public = false
# file_size_limit = "50MiB"
# allowed_mime_types = ["image/png", "image/jpeg"]
# objects_path = "./images"

[auth]
enabled = true
# The base URL of your website. Used as an allow-list for redirects and for constructing URLs used
# in emails.
site_url = "http://127.0.0.1:3000"
# A list of *exact* URLs that auth providers are permitted to redirect to post authentication.
additional_redirect_urls = ["https://127.0.0.1:3000"]
# How long tokens are valid for, in seconds. Defaults to 3600 (1 hour), maximum 604,800 (1 week).
jwt_expiry = 3600
# If disabled, the refresh token will never expire.
enable_refresh_token_rotation = true
# Allows refresh tokens to be reused after expiry, up to the specified interval in seconds.
# Requires enable_refresh_token_rotation = true.
refresh_token_reuse_interval = 10
# Allow/disallow new user signups to your project.
enable_signup = true
# Allow/disallow anonymous sign-ins to your project.
enable_anonymous_sign_ins = false
# Allow/disallow testing manual linking of accounts
enable_manual_linking = false

[auth.email]
# Allow/disallow new user signups via email to your project.
enable_signup = true
# If enabled, a user will be required to confirm any email change on both the old, and new email
# addresses. If disabled, only the new email is required to confirm.
double_confirm_changes = true
# If enabled, users need to confirm their email address before signing in.
enable_confirmations = false
# If enabled, users will need to reauthenticate or have logged in recently to change their password.
secure_password_change = false
# Controls the minimum amount of time that must pass before sending another signup confirmation or password reset email.
max_frequency = "1s"

# Use a production-ready SMTP server
# [auth.email.smtp]
# host = "smtp.sendgrid.net"
# port = 587
# user = "apikey"
# pass = "env(SENDGRID_API_KEY)"
# admin_email = "admin@email.com"
# sender_name = "Admin"

# Uncomment to customize email template
# [auth.email.template.invite]
# subject = "You have been invited"
# content_path = "./supabase/templates/invite.html"

[auth.sms]
# Allow/disallow new user signups via SMS to your project.
enable_signup = true
# If enabled, users need to confirm their phone number before signing in.
enable_confirmations = false
# Template for sending OTP to users
template = "Your code is {{ .Code }} ."
# Controls the minimum amount of time that must pass before sending another sms otp.
max_frequency = "5s"

# Use pre-defined map of phone number to OTP for testing.
# [auth.sms.test_otp]
# 4152127777 = "123456"

# Configure logged in session timeouts.
# [auth.sessions]
# Force log out after the specified duration.
# timebox = "24h"
# Force log out if the user has been inactive longer than the specified duration.
# inactivity_timeout = "8h"

# This hook runs before a token is issued and allows you to add additional claims based on the authentication method used.
# [auth.hook.custom_access_token]
# enabled = true
# uri = "pg-functions://<database>/<schema>/<hook_name>"

# Configure one of the supported SMS providers: `twilio`, `twilio_verify`, `messagebird`, `textlocal`, `vonage`.
[auth.sms.twilio]
enabled = false
account_sid = ""
message_service_sid = ""
# DO NOT commit your Twilio auth token to git. Use environment variable substitution instead:
auth_token = "env(SUPABASE_AUTH_SMS_TWILIO_AUTH_TOKEN)"

[auth.mfa]
# Control how many MFA factors can be enrolled at once per user.
max_enrolled_factors = 10

# Control use of MFA via App Authenticator (TOTP)
[auth.mfa.totp]
enroll_enabled = true
verify_enabled = true

# Configure Multi-factor-authentication via Phone Messaging
# [auth.mfa.phone]
# enroll_enabled = true
# verify_enabled = true
# otp_length = 6
# template = "Your code is {{ .Code }} ."
# max_frequency = "10s"

# Use an external OAuth provider. The full list of providers are: `apple`, `azure`, `bitbucket`,
# `discord`, `facebook`, `github`, `gitlab`, `google`, `keycloak`, `linkedin_oidc`, `notion`, `twitch`,
# `twitter`, `slack`, `spotify`, `workos`, `zoom`.
[auth.external.apple]
enabled = false
client_id = ""
# DO NOT commit your OAuth provider secret to git. Use environment variable substitution instead:
secret = "env(SUPABASE_AUTH_EXTERNAL_APPLE_SECRET)"
# Overrides the default auth redirectUrl.
redirect_uri = ""
# Overrides the default auth provider URL. Used to support self-hosted gitlab, single-tenant Azure,
# or any other third-party OIDC providers.
url = ""
# If enabled, the nonce check will be skipped. Required for local sign in with Google auth.
skip_nonce_check = false

# Use Firebase Auth as a third-party provider alongside Supabase Auth.
[auth.third_party.firebase]
enabled = false
# project_id = "my-firebase-project"

# Use Auth0 as a third-party provider alongside Supabase Auth.
[auth.third_party.auth0]
enabled = false
# tenant = "my-auth0-tenant"
# tenant_region = "us"

# Use AWS Cognito (Amplify) as a third-party provider alongside Supabase Auth.
[auth.third_party.aws_cognito]
enabled = false
# user_pool_id = "my-user-pool-id"
# user_pool_region = "us-east-1"

[edge_runtime]
enabled = true
# Configure one of the supported request policies: `oneshot`, `per_worker`.
# Use `oneshot` for hot reload, or `per_worker` for load testing.
policy = "oneshot"
inspector_port = 8083

[analytics]
enabled = true
port = 64327
# Configure one of the supported backends: `postgres`, `bigquery`.
backend = "postgres"

# Experimental features may be deprecated any time
[experimental]
# Configures Postgres storage engine to use OrioleDB (S3)
orioledb_version = ""
# Configures S3 bucket URL, eg. <bucket_name>.s3-<region>.amazonaws.com
s3_host = "env(S3_HOST)"
# Configures S3 bucket region, eg. us-east-1
s3_region = "env(S3_REGION)"
# Configures AWS_ACCESS_KEY_ID for S3 bucket
s3_access_key = "env(S3_ACCESS_KEY)"
# Configures AWS_SECRET_ACCESS_KEY for S3 bucket
s3_secret_key = "env(S3_SECRET_KEY)"
```

Error: supabase_vector_fiji-tax-pro container is not ready: unhealthy

**Comment by znowfox** [comment]
what you really want to do is run seperate networks on each of these services, keep config pretty much default

**Comment by 80Builder80** [comment]
@code-leen could you expand please? I need some clarification on this approach. 
In order to have multiple projects for locally hosted Supabase, does the entire Supabase stack have to be created for each project? 
When deploying Supabase via Docker, 13 containers are created (studio, auth, realtime, kong, vector, etc.). 
Every project will require 13 containers? 
Are there services that are not required and can be eliminated from the service stack? 
Since this is a 5 month old discussion, has this approach changed?

My primary use for Supabase at this time is for agent vector documentation retrieval via N8N. I'm sure I will have other uses for Supabase in the future. However, I do not intent to place Supabase in production use. It will solely be used in my local environment for personal use.

**Reply by znowfox** [reply]
So yeah full services stack every instances, it's bad, it's heavy, it's buggy, it's feature missing. They do it on purpose, what i can advise to you for dev purposes is to do one email one online acc.

**Comment by guyrandalf** [comment]
@code-leen you are absolutely right and this works for me. One issue with this local setup is some things still break in the studio like accessing the /api route. This way if you fail to copy the credentials from the terminal initially into the project, you will never see the details again which is not how the cloud works.

The idea which is pretty bad, is to have a single instance per project. And then go into the toml file and update the port

Anyway, thanks

**Comment by LambdaSoftworks** [comment]
I built a program to manage the deployment of multiple self hosted instances using docker. My contact info is in my profile, reach out if anyone is interested in getting a copy.

**Comment by taimoorkhokhar** [comment]
Anyone interested here to contribute to create a fork for this solution? would love to collaborate. Since so many people need this solution.

**Reply by LambdaSoftworks** [reply]
Check out www.supascale.io

It's still in development, but the core feature you're after works. I use it for 3 production databases.

**Reply by tothambrus11** [reply]
Careful, that link appears to be refurbished for something else. May be a scam.
